Cabal-tests/lib/Test/Utils/TempTestDir.hs

```diff
@@ -23,12 +24,17 @@ import qualified System.Info (os)
 -- | Much like 'withTemporaryDirectory' but with a number of hacks to make
 -- sure on windows that we can clean up the directory at the end.
 withTestDir :: (MonadIO m, MonadMask m) => Verbosity -> String -> (FilePath -> m a) -> m a
-withTestDir verbosity template action = do
+withTestDir verbosity template action = withTestDir' verbosity defaultTempFileOptions template action
+
+withTestDir' :: (MonadIO m, MonadMask m) => Verbosity -> TempFileOptions -> String -> (FilePath -> m a) -> m a
+withTestDir' verbosity tempFileOpts template action = do
   systmpdir <- liftIO getTemporaryDirectory
   bracket
     ( do { tmpRelDir <- liftIO $ createTempDirectory systmpdir template
          ; return $ systmpdir </> tmpRelDir } )
-    (liftIO . removeDirectoryRecursiveHack verbosity)
+    -- Windows sometimes fails to remove the directory if there was a s-r-p.
+    -- This alleviates the problem.
+    (liftIO . when (not (optKeepTempFiles tempFileOpts)) . removeDirectoryRecursiveHack verbosity)
```

Comment:
The comment doesn't describe the added `when` does it? It just explains why `removeDIrectoryRecursiveHack` is there? But then I'm not sure the comment adds anything to the top-level comment for `withTempDir`... Or am I missing something?

Reply:
The `when` comes from the change I made for tests to use this instead of `withTempDirectoryEx`. It is the key for the tests to not delete their temporal directory.

Probably I should write that down in a comment. Also I missed the other comment you mention, what I wrote seems indeed redundant.

Reply:
Good, that's what I thought. Yeah, adding a comment about temp dirs and (re)moving duplication would be great.

Reply:
Done